File: modules/data/mp3/mp3.js
/*
 * Copyright (c) 2023  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Runtime.
 * 
 *   The Moddable SDK Runtime is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 * 
 *   The Moddable SDK Runtime is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 * 
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with the Moddable SDK Runtime.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

// tables based on https://github.com/biril/mp3-parser/blob/master/lib/lib.js
// http://mpgedit.org/mpgedit/mpeg_format/mpeghdr.htm
// https://github.com/lieff/minimp3/blob/master/minimp3.h#L77

 const bitRates = [
	[],
	[],
	[
		null,
		[0, 8, 16, 24, 32, 40, 48, 56, 64, 80, 96, 112, 128, 144, 160],
		[0, 8, 16, 24, 32, 40, 48, 56, 64, 80, 96, 112, 128, 144, 160],
		[0, 32, 48, 56, 64, 80, 96, 112, 128, 144, 160, 176, 192, 224, 256]
	],
	[
		null,
		[0, 32, 40, 48, 56, 64, 80, 96, 112, 128, 160, 192, 224, 256, 320],
		[0, 32, 48, 56, 64, 80, 96, 112, 128, 160, 192, 224, 256, 320, 384],
		[0, 32, 64, 96, 128, 160, 192, 224, 256, 288, 320, 352, 384, 416, 448]
	],
];
bitRates[0] = bitRates[2];
Object.freeze(bitRates, true);

const sampleRates = Object.freeze([
	[11025, 12000, 8000],
	[],
	[22050, 24000, 16000],
	[44100, 48000, 32000]
], true);

const sampleLengths = Object.freeze([
	[],
	[576, 1152, 384],
	[576, 1152, 384],
	[1152, 1152, 384],
], true);

export default class @ "xs_mp3_destructor" {
	constructor() @ "xs_mp3";
	close() @ "xs_mp3_close";
	decode(input, output) @ "xs_mp3_decode";
	
	static scan(buffer, start, end, info = {}) {
		let position = sync(buffer, start, end, info);
		if (undefined === position)
			return;

		info.position = position;

		if (0 === info.length) {
			position = sync(buffer, position + 1, end);
			if ((undefined === position) && ((end - start) >= 2048))
				return;

			info.length = 2048;		// worst case (MAINBUF_SIZE == 1940 in libhelix)
		}

		return info;
	}
	static BUFFER_GUARD = 8;		// MAD_BUFFER_GUARD == 8
}

function sync(buffer, position, end, info)
{
	while (position < end) {
		position = buffer.indexOf(0xFF, position);
		if ((position < 0) || ((position + 4) > end))
			return;

		let byte = buffer[position + 1];
		if (0xE0 !== (byte & 0xE0)) {
			position += 1;
			continue;
		}

		const mpegVersion = (byte >> 3) & 0x03;
		const layerVersion = (byte >> 1) & 0x03;
		if ((1 === mpegVersion) || (0 === layerVersion)) {
			position += 1;
			continue;
		}

		byte = buffer[position + 2];
		const bitRate = bitRates[mpegVersion][layerVersion][byte >> 4];
		const sampleRate = sampleRates[mpegVersion][(byte >> 2) & 3];
		const padded = (byte >> 1) & 1;
		if ((undefined === bitRate) || (undefined === sampleRate)) {
			position += 1;
			continue;
		}

		if (info) {
			let length = sampleLengths[mpegVersion][layerVersion];
			if (length) {
				length = Math.idiv(length * (bitRate * 1000 / 8), sampleRate);
				if (padded)
					length += (3 === layerVersion) ? 4 : 1;
			}
			info.length = length;		// will be 0 if unable to calculate it here... for LFR may be too big... that's OK too
		}
		
		return position;
	}
}

File: modules/data/mp3/mp3.c
/*
 * Copyright (c) 2023  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Runtime.
 * 
 *   The Moddable SDK Runtime is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 * 
 *   The Moddable SDK Runtime is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 * 
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with the Moddable SDK Runtime.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "xsHost.h"
#include "xsmc.h"
#include "mc.xs.h"

#include "./libmad/config.h"
#include "./libmad/mad.h"

#define kMP3SamplesPerFrame (1152)

struct xsMP3Record {
	struct mad_stream stream;
	struct mad_frame frame;
	struct mad_synth synth;

	uint8_t *outputBuffer;
	int32_t outputBufferPosition;
};
typedef struct xsMP3Record xsMP3Record;
typedef struct xsMP3Record *xsMP3;

void xs_mp3_destructor(void *data)
{
	xsMP3 mp3 = data;
	if (!mp3) return;

	mad_synth_finish(&mp3->synth);
	mad_frame_finish(&mp3->frame);
	mad_stream_finish(&mp3->stream);
	c_free(mp3);
}

void xs_mp3(xsMachine *the)
{
	xsMP3 mp3;
	
	mp3 = c_calloc(1, sizeof(xsMP3Record));
	if (!mp3) xsUnknownError("no memory");

	xsmcSetHostData(xsThis, mp3);

	mad_stream_init(&mp3->stream);
	mad_frame_init(&mp3->frame);
	mad_synth_init(&mp3->synth);
}

void xs_mp3_close(xsMachine *the)
{
	xsMP3 mp3 = xsmcGetHostData(xsThis);
	xs_mp3_destructor(mp3);
	xsmcSetHostData(xsThis, NULL);
}

static enum mad_flow output(void *s, struct mad_header const *header, struct mad_pcm *pcm)
{
	xsMP3 mp3 = s;
	if (MAD_MODE_SINGLE_CHANNEL == header->mode)
		c_memcpy(mp3->outputBuffer + mp3->outputBufferPosition, pcm->samples[0], pcm->length * 2);
	else {
		int16_t *left = (int16_t *)pcm->samples[0];
		int16_t *right = (int16_t *)pcm->samples[1];
		int16_t *out = (int16_t *)(mp3->outputBuffer + mp3->outputBufferPosition);
		int count = pcm->length;
		while (count--)
			*out++ = (*left++ + *right++) >> 1;
	}
	mp3->outputBufferPosition += pcm->length * 2;

	return MAD_FLOW_CONTINUE;
}

void xs_mp3_decode(xsMachine *the)
{
	xsMP3 mp3 = xsmcGetHostData(xsThis);
	void *data;
	xsUnsignedValue count;
	
	if (!mp3)
		xsUnknownError("closed");

	xsmcGetBufferWritable(xsArg(1), (void *)&mp3->outputBuffer, &count);
	if ((kMP3SamplesPerFrame * sizeof(int16_t)) > count)
		xsUnknownError("output too small");
	mp3->outputBufferPosition = 0;

	xsmcGetBufferReadable(xsArg(0), &data, &count);
	mad_stream_buffer(&mp3->stream, data, count);		// count includes MAD_BUFFER_GUARD

	mad_frame_decode(&mp3->frame, &mp3->stream);

	enum mad_flow result = mad_synth_frame(&mp3->synth, &mp3->frame, output, mp3);
	if (MAD_FLOW_BREAK == result)
		xsUnknownError("break");

	if (!mp3->stream.this_frame || !mp3->stream.next_frame)
		return;

	xsmcSetInteger(xsResult, mp3->outputBufferPosition / sizeof(int16_t));
	xsmcSet(xsArg(1), xsID_samples, xsResult);
	xsmcSetInteger(xsResult, mp3->stream.next_frame - mp3->stream.this_frame);
}
